File: HaishinKit.dart/android/src/main/kotlin/com/haishinkit/haishin_kit/RtmpConnectionHandler.kt
package com.haishinkit.haishin_kit

import com.haishinkit.rtmp.RtmpConnection
import com.haishinkit.rtmp.event.Event
import com.haishinkit.rtmp.event.IEventListener
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class RtmpConnectionHandler(
    private val plugin: HaishinKitPlugin
) : MethodChannel.MethodCallHandler, IEventListener, EventChannel.StreamHandler {
    companion object {
        private const val TAG = "RtmpConnection"
    }

    var instance: RtmpConnection? = RtmpConnection()
        private set(value) {
            field?.dispose()
            field = value
        }

    private var channel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
        set(value) {
            field?.endOfStream()
            field = value
        }

    init {
        instance?.addEventListener(Event.RTMP_STATUS, this)
        channel = EventChannel(
            plugin.flutterPluginBinding.binaryMessenger, "com.haishinkit.eventchannel/${hashCode()}"
        )
        channel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "$TAG#connect" -> {
                val command = call.argument<String>("command") ?: ""
                instance?.connect(command)
                result.success(null)
            }

            "$TAG#close" -> {
                instance?.close()
                result.success(null)
            }

            "$TAG#dispose" -> {
                eventSink = null
                instance = null
                plugin.onDispose(hashCode())
                result.success(null)
            }
        }
    }

    override fun handleEvent(event: Event) {
        val map = HashMap<String, Any?>()
        map["type"] = event.type
        map["data"] = event.data
        plugin.uiThreadHandler.post {
            eventSink?.success(map)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
    }
}

File: HaishinKit.dart/ios/haishin_kit/Sources/RTMPConnectionHandler.swift
import Foundation
#if canImport(Flutter)
import Flutter
#endif
#if canImport(FlutterMacOS)
import FlutterMacOS
#endif
import HaishinKit
import RTMPHaishinKit

final class RTMPConnectionHandler: NSObject, MethodCallHandler {
    var instance: RTMPConnection?
    private let plugin: HaishinKitPlugin
    private var channel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var subscription: Task<(), Error>? {
        didSet {
            oldValue?.cancel()
        }
    }

    init(plugin: HaishinKitPlugin) {
        self.plugin = plugin
        super.init()
        let id = Int(bitPattern: ObjectIdentifier(self))
        if let messanger = plugin.registrar?.messenger() {
            self.channel = FlutterEventChannel(name: "com.haishinkit.eventchannel/\(id)", binaryMessenger: messanger)
        } else {
            self.channel = nil
        }
        instance = RTMPConnection()
        channel?.setStreamHandler(self)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "RtmpConnection#connect":
            guard
                let arguments = call.arguments as? [String: Any?],
                let command = arguments["command"] as? String else {
                return
            }
            if let instance {
                subscription = Task { [weak self] in
                    for await status in await instance.status {
                        DispatchQueue.main.async { [eventSink = self?.eventSink] in
                            eventSink?(status.makeEvent())
                        }
                    }
                }
            }
            Task {
                do {
                    _ = try await instance?.connect(command)
                } catch {
                    result(FlutterError(code: "RTMP_CONNECT_FAILED", message: String(describing: error), details: nil))
                    return
                }
                result(nil)
            }
        case "RtmpConnection#close":
            subscription = nil
            Task {
                try? await instance?.close()
                result(nil)
            }
        case "RtmpConnection#dispose":
            instance = nil
            plugin.onDispose(id: Int(bitPattern: ObjectIdentifier(self)))
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension RTMPConnectionHandler: FlutterStreamHandler {
    // MARK: FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}
